BootloaderCommonPkg/Library/ElfLib/ElfLib.c

```diff
@@ -0,0 +1,228 @@
+
+    CopyMem ((VOID *)Addr, (VOID *) ((UINT32)ImageBase + Shdr[Index].Sh_Offset), Size);
+    Shdr[Index].Sh_Addr = Addr;
+    Eaddr = Addr + Size;
+  }
+
+  return Eaddr;
+}
+
+/**
+  Load the ELF image to specified address in ELF header.
+
+  This function load ELF image section by section into memory address specified
+  in ELF program header, and also load ELF symbols.
+
+  @param[in]  ImageBase      Memory address of an image
+
+  @retval Image entry point  The entry point of ELF image if load image success
+  @retval NULL               Error with loading ELF image
+**/
+UINT32 *
+LoadElfImage (
+  IN  VOID  *ImageBase
+  )
+{
+  ELF32_EHDR                 *Ehdr;
+  ELF32_PHDR                 *Phdr;
+  UINT32                      Eaddr;
+  UINT32                      Index;
+  UINT32                      Msize;
+  UINT32                      Fsize;
+  UINT32                      Mpos;
+  UINT32                      Ipos;
+
+  //
+  // TODO: Check if image, load region overlap; if yes, move image away.
+  //
+
+  //
+  // Get ELF header and Program header
+  //
+  Ehdr = (ELF32_EHDR *) ImageBase;
+  Phdr = (ELF32_PHDR *) ((UINT32)ImageBase + Ehdr->E_Phoff);
+
+  //
+  //  Iterate over the program headers and load all necessary segments.
+  //  Keep track of the highest load address used in <eaddr>
+  //
+  Eaddr = 0;
+  for (Index = 0; Index < Ehdr->E_Phnum ; Index ++) {
+    if (Phdr[Index].P_Type != PT_LOAD) {
+      continue;
+    }
+
+    Msize = Phdr[Index].P_Memsz;            // target (in-memory) segment size
+    Fsize = Phdr[Index].P_Filesz;           // source (image) segment size
+    Mpos  = Phdr[Index].P_Paddr;            // target segment address
+    Ipos  = Phdr[Index].P_Offset;           // location of segment in image
+
+    if ((Mpos <= (UINT32)ImageBase) && (Mpos + Msize > (UINT32)ImageBase)) {
+      DEBUG ((DEBUG_INFO, "** IMAGE overwrite, out of range **\n"));
+      return NULL;
+    }
+
+    if (Fsize != 0) {
+      CopyMem ((VOID *) Mpos, (VOID *) ((UINT32)ImageBase + Ipos), Fsize);
+    }
+
+    if (Msize > Fsize) {
+      ZeroMem ((VOID *) (Mpos + Fsize), Msize - Fsize);
+    }
+
+    Eaddr = MAX (Eaddr, Mpos + Msize);
+  }
+  ASSERT (Eaddr != 0);
+
+  //
+  // The image is now bootable; try to load the symbol table as well
+  //
+  LoadElfSymtab (ImageBase, ALIGN_UP (Eaddr, SIZE_4KB));
+
+  //
+  // Return Entry Point
+  //
+  return (UINT32 *) Ehdr->E_Entry;
+}
+
+/**
+  Update MultibootInfo from ELF
+
+  This function updates Multiboot Information if there is symbol information in ELF.
+  Caller must guarantee the image is ELF, multiboot and loaded successfully.
+
+  @param[in]        ImageBase   Memory address of an image
+  @param[in, out]   MbInfo      Pointer to MULTIBOOT_INFO which needs to be updated
+
+  @retval RETURN_SUCCESS        MULTIBOOT_INFO successfully updated.
+  @retval RETURN_NOT_FOUND      No ELF Symbol information found.
+
+**/
+RETURN_STATUS
+UpdateElfMultibootInfo (
```

Comment:
I think we should try to move MULTBOOT related code into Multiboot library.  Here if required, create new function to return ELF info to be populated into MULTIBOOT_INFO.

Reply:
Okay, let me create a new function for ELF info and use it in MultibootLib.